core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -290,12 +336,24 @@ public void requestConnection() {
   }
 
   private Subchannel createNewSubchannel(SocketAddress addr) {
+    SubchannelStateListener hcListener = new SubchannelStateListener() {
+      @Override
+      public void onSubchannelState(ConnectivityStateInfo newState) {
+        onSubchannelHealth(this, newState);
```

Comment:
It's unfortunate that we can no longer create the `SubchannelData` before the listener. Instead of linear scanning the subchannels, would it be nicer to do (the last line is the most important one):

```java
  class HealthListener implements SubchannelStateListener {
      private SubchannelData subchannelData;
      @Override
      public void onSubchannelState(ConnectivityStateInfo newState) {
        // Consider what to do on null subchannelData here. You could return early, but it may be better to make it
        // break (throw an exception).
        onSubchannelHealth(subchannelData, newState);
      }
  }
  HealthListener hcListener = new HealthListener();
  ...
  SubchannelData subchannelData = new SubchannelData(subchannel, IDLE, hcListener);
  hcListener.subchannelData = subchannelData;
```

If you store the state in the health listener (instead of the `SubchannelData`), you can make it a bit nicer because it doesn't need to do any lookups to update the state:
```java
  class HealthListener implements SubchannelStateListener {
      private ConnectivityStateInfo healthState = ConnectivityStateInfo.forNonError(IDLE);

      @Override
      public void onSubchannelState(ConnectivityStateInfo newState) {
        log.log(Level.FINE, "Received health status {0} for listener {1}",
            new Object[]{newState, this});
        this.healthState = newState;
        if (addressIndex.isValid()
              && subchannels.get(addressIndex.getCurrentAddress()).getHealthListener() == this) {
          updateHealthCheckedState(subchannels.get(addressIndex.getCurrentAddress()));
        }
      }
  }
```

Reply:
yea that should have been fixed..